style: standardise prettier config and add to pre-commit commit hook

toddym42 opened the pull request:
## Introduction

Prettier is an opinionated code formatter that can help keep a consistent code formatting style in the project. It has been part implemented in the project but there are the following issues:

* The `prettier` package and config have been added in the root project, but some individual sub-projects (`external-api` and `portal-api`) have their own version of `prettier` and config added. This may lead to inconsistencies if the versions and config diverge
* Sometimes running prettier on files results in conflicts with the existing eslint config, so prettier changes need to be reversed
* Running of prettier is not enforced in any way (either with commit hooks or CI) so there is no prompt to the dev that a file has not been formatted

## Resolution

* Remove the `prettier` package and config from sub-projects so that formatting across the entire codebase will be based on the root directory prettier installation and config. Updated the README to clarify this point.
* Added [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) package to all projects - this resolves any conflicts between eslint and prettier so we can ensure that prettier can always be run
* Added running of prettier to the commit hook so that this is always run on staged files even if the dev hasn't run the formatter manually
* Moved the running of `lint-staged` to a "pre-commit" hook. This means that any eslint or prettier changes that get done as part of the commit hook are done before the commit is submitted, meaning they will be included in commit you are submitting. Previously this was included in the "commit-msg" hook, meaning you would add a commit, then need to add a further separate commit with any eslint or prettier changes. Further details of the different types of commit hooks [here](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)

JoshBinns2000 reviewed (commented):
Looks good! Very happy we're standardising this finally. Is there any way to automatically update all the existing files to match the lint as part of this PR so that future PRs don't have excessive file updates which fix existing linting errors? Also, is there any way that we can enforce the `.eslintrc.js` files across directories being the same so that people don't change them by accident?

toddym42 commented on external-api/package.json:
```diff
@@ -68,10 +68,9 @@
     "eslint-import-resolver-typescript": "^3.6.1",
     "eslint-plugin-import": "^2.29.0",
-    "eslint-plugin-prettier": "4.2.1",
+    "eslint-plugin-prettier": "^5.0.1",
```
I haven't actually added any `.eslintrc.js` files - all the projects had them already and there are a lot of inconsistencies. Maybe as a separate followup exercise we can standardise these as well at some point.

toddym42 commented:
> Looks good! Very happy we're standardising this finally. Is there any way to automatically update all the existing files to match the lint as part of this PR so that future PRs don't have excessive file updates which fix existing linting errors? Also, is there any way that we can enforce the `.eslintrc.js` files across directories being the same so that people don't change them by accident?

@JoshBinns2000

Note that these changes are primarily focussed on prettier formatting, rather than the full scope of what eslint analyses. While prettier focusses only on formatting, eslint has a wider set of concerns including logic errors, best practices, and formatting. It's because eslint also includes formatting that I added the [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) package to prevent conflicts between eslint and prettier. Note that there are plans to remove all formatting rules in future eslint releases to prevent these kinds of conflicts in future (see [here](https://eslint.org/blog/2023/10/deprecating-formatting-rules/) for details).

As part of these changes I could have also run prettier across the whole project to get everything in line, but this would have touched thousands of files and would have caused merge conflict nightmares for everyone currently working on feature branches. We might make the call to do this at some point and take the hit, but we thought best to hold off for now.

With prettier now added to the `pre-commit` commit hook, this will only affect files that have been changed as part of your commit, so we will be incrementally getting everything in line.

Happy to have a call if you want to discuss any of this further - let me know.

toddym42 commented on .prettierignore:
```diff
@@ -1,6 +1,4 @@
+dist
+.cache
+coverage
+node_modules
```
If we just let the commit hook do its thing then we are only running on `.js` and `.ts` files, as that's what's defined in the `lint-staged` section on the root `package.json`
```
"**/*.{js,ts}": [
  "eslint --fix",
  "prettier --write"
],
```
I'm imagining when we get to the point that we'd be happy to run prettier on the entire codebase we'd add the following extra `package.json` scripts:
```
"prettier": "prettier --write \"./**/*.{ts,js}\"",
"prettier:check": "prettier --check \"./**/*.{ts,js}\"",
```
and we could add `prettier:check` to the existing `lint:all` script.

Maybe safer to just leave `package-lock.json` just in case though. I'll re-add.

lmarrai-sw replied:
Ah I see, I didn't think of that. I'm happy then, thanks @toddym42
